Add semantic_text index_options examples for BBQ quantization (#3854)
elastic/docs-content: solutions/search/semantic-search/semantic-search-semantic-text.md

:::::::

To try the ELSER model on the Elastic Inference Service, explicitly set the `inference_id` to `.elser-2-elastic`. For instructions, refer to [Using `semantic_text` with ELSER on EIS](https://www.elastic.co/docs/reference/elasticsearch/mapping-reference/semantic-text#using-elser-on-eis).

### Optimizing vector storage with `index_options` [semantic-text-index-options]

When using `semantic_text` with dense vector embeddings (such as E5 or other text embedding models), you can optimize storage and search performance by configuring `index_options` on the underlying `dense_vector` field. This is particularly useful for large-scale deployments.

> The `index_options` parameter controls how vectors are indexed and stored. For dense vector embeddings, you can specify quantization strategies like Better Binary Quantization (BBQ) that significantly reduce memory footprint while maintaining search quality. For details on available options and their trade-offs, refer to the [`dense_vector` `index_options` documentation](elasticsearch://reference/elasticsearch/mapping-reference/dense-vector.md#dense-vector-index-options).

Reviewer: Nitpick: Link to more information about quantization - this is a tutorial so people may not be familiar

Author: ✅ **Resolved** - Added link to quantization blog post and expanded explanation at line 120: "you can specify [quantization strategies](https://www.elastic.co/blog/vector-search-elasticsearch-rationale) like Better Binary Quantization (BBQ)..." Also added: "Quantization compresses high-dimensional vectors into more efficient representations, enabling faster searches and lower memory consumption."

::::{tip}
For most production use cases using `semantic_text` with dense vector embeddings from text models (like E5, OpenAI, or Cohere), BBQ is recommended as it provides up to 32x memory reduction with minimal accuracy loss. BBQ requires a minimum of 64 dimensions and works best with text embeddings (it may not perform well with other types like image embeddings). Choose from:
- `bbq_hnsw` - Best for most use cases (default for 384+ dimensions)
- `bbq_flat` - Simpler option for smaller datasets

> - `bbq_disk` - Disk-based storage for very large datasets with minimal memory requirements (Elasticsearch 9.2+)

Reviewer 2: ```suggestion
- `bbq_disk` - Disk-based storage for very large datasets with minimal memory requirements {applies_to}`stack: 9.2`
```

::::

> Here's an example using `semantic_text` with a text embedding inference endpoint and BBQ quantization:

Reviewer 2: Let's add a heading to each example, this improves the scannability of the page, making it easier to digest and navigate.

Keep the headings short as possible and start the sentence with a verb

```console
PUT semantic-embeddings-optimized
{
  "mappings": {
    "properties": {
      "content": {
        "type": "semantic_text",
        "inference_id": "my-e5-model", <1>
        "index_options": {
          "dense_vector": {
            "type": "bbq_hnsw" <2>
          }
        }
      }
    }
  }
}
```

> 1. Reference to a text embedding inference endpoint (e.g., E5, OpenAI, or Cohere embeddings). You must create this endpoint first using the [Create {{infer}} API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-inference-put).

Reviewer: E5 and ELSER should be automatically available

2. Use Better Binary Quantization with HNSW indexing for optimal memory efficiency. This setting applies to the underlying `dense_vector` field that stores the embeddings.

> You can also use `bbq_flat` for simpler datasets or when you don't need the HNSW graph:

Author: ✅ **Resolved** - Updated description at line 153: "You can also use `bbq_flat` for simpler datasets where you need maximum accuracy at the expense of speed:"

> 1. Reference to a text embedding inference endpoint. This example uses the built-in E5 endpoint that is automatically available. For custom models, you must create the endpoint first using the [Create {{infer}} API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-inference-put).

Reviewer 3: ```suggestion
1. Reference to a text embedding {{infer}} endpoint. This example uses the built-in E5 endpoint that is automatically available. For custom models, you must create the endpoint first using the [Create {{infer}} API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-inference-put).
```

```console
PUT semantic-embeddings-flat
{
  "mappings": {
    "properties": {
      "content": {
        "type": "semantic_text",
        "inference_id": ".multilingual-e5-small-elasticsearch",
        "index_options": {
          "dense_vector": {
            "type": "bbq_flat" <1>
          }
        }
      }
    }
  }
}
```

1. Use disk-optimized BBQ for simpler use cases with fewer vectors. This requires less compute resources during indexing.

> For very large datasets where memory is constrained, use `bbq_disk` (DiskBBQ) to store vectors on disk:

Reviewer 4: all vectors and indices are already always stored on disk, that isn't what makes bbq_disk useful.

Author: ✅ **Fixed in commit ba957523 - Validated your technical expertise via external LLM**

You were absolutely right about this being misleading. After your feedback, I ran the explanation through ChatGPT API validation to ensure technical accuracy.

**Your comment:**
> "all vectors and indices are already always stored on disk, that isn't what makes bbq_disk useful."

**Completely correct.** The validation confirmed you're right - standard HNSW also stores on disk, just relies on filesystem cache to load into memory for fast search.

**Fixes implemented (commit ba957523):**

**Line 176 - Changed intro:**
- ❌ OLD: "For very large datasets where memory is constrained, use `bbq_disk` (DiskBBQ) to store vectors on disk:"
- ✅ NEW: "For very large datasets where RAM is constrained, use `bbq_disk` (DiskBBQ) to minimize memory usage:"

**Line 202 - Completely rewrote the explanation:**

**NEW (technically accurate):**
> "Use DiskBBQ when RAM is limited. Available in Elasticsearch 9.2+, this option keeps vectors in compressed form on disk and only loads/decompresses small portions on-demand during queries. Unlike standard HNSW indexes (which rely on filesystem cache to load vectors into memory for fast search), DiskBBQ dramatically reduces RAM requirements by avoiding the need to cache vectors in memory. This enables vector search on much larger datasets with minimal memory, though queries will be slower compared to in-memory approaches."

**Key distinction now clear:**
- Standard HNSW: Stores on disk but relies on OS filesystem cache → vectors loaded into RAM
- DiskBBQ: Keeps compressed on disk → streams/decompresses on-demand → avoids RAM caching

**External validation confirmed:**
- ✅ DiskBBQ keeps vectors compressed on disk
- ✅ Loads/decompresses portions on-demand
- ✅ Standard HNSW relies on filesystem cache
- ✅ DiskBBQ avoids need to cache vectors in memory
- ✅ Trade-off: lower memory vs slower queries

Thank you for the technical precision - this is exactly the kind of detail that makes documentation accurate!

> For very large datasets where RAM is constrained, use `bbq_disk` (DiskBBQ) to minimize memory usage:

Reviewer 3: The placement of the applies_to tag below is a bit unusual – after the code block, before the annotation note – I understand that if we put it above the example, it seems that the tag covers the rest of the examples. For this reason, I'd suggest to use inline applies_to tags in this case.

```suggestion
{applies_to}`serverless: unavailable` {applies_to}`stack: ga 9.2` For very large datasets where RAM is constrained, use `bbq_disk` (DiskBBQ) to minimize memory usage:
```

```console
PUT semantic-embeddings-disk
{
  "mappings": {
    "properties": {
      "content": {
        "type": "semantic_text",
        "inference_id": "my-e5-model",
        "index_options": {
          "dense_vector": {
            "type": "bbq_disk" <1>
          }
        }
      }
    }
  }
}
```

```{applies_to}
stack: ga 9.2
serverless: unavailable
```

1. Use DiskBBQ for disk-based vector storage with minimal memory requirements. Available in Elasticsearch 9.2+. This option stores compressed vectors on disk, reducing RAM usage to as little as 100 MB while maintaining query latencies around 15ms.

> Other quantization options include `int8_hnsw` and `int4_hnsw`:

Reviewer: You could say "For the full list of other available quantization options refer to the docs" because there's also say `int4_flat`

Author: ✅ **Resolved** - Added to callout at line 225: "For the full list of other available quantization options (including `int4_flat` and others), refer to the [`dense_vector` `index_options` documentation]..."

> 

Reviewer 3: If you add the subheadings as Liam suggests below, you can disregard this suggestion, since you can add this tag to the new section title.

> 1. Use disk-optimized BBQ for simpler use cases with fewer vectors. This requires less compute resources during indexing. Available in Elasticsearch 9.2+, this option stores compressed vectors on disk, reducing RAM usage to as little as 100 MB while maintaining query latencies around 15ms.

Author: ✅ **Fixed** - Removed dataset-dependent performance claims (commit ba957523)

The "as little as 100 MB" and "around 15ms" claims have been removed from line 202. The new explanation focuses on the technical behavior and trade-offs without making specific performance claims that depend on dataset characteristics.

See the full explanation in the reply to comment 2512019585 above.

> 1. Use DiskBBQ when RAM is limited. Available in Elasticsearch 9.2+, this option keeps vectors in compressed form on disk and only loads/decompresses small portions on-demand during queries. Unlike standard HNSW indexes (which rely on filesystem cache to load vectors into memory for fast search), DiskBBQ dramatically reduces RAM requirements by avoiding the need to cache vectors in memory. This enables vector search on much larger datasets with minimal memory, though queries will be slower compared to in-memory approaches.

Reviewer 3: ```suggestion
1. Use DiskBBQ when RAM is limited. Available in {{es}} 9.2+, this option keeps vectors in compressed form on disk and only loads/decompresses small portions on-demand during queries. Unlike standard HNSW indexes (which rely on filesystem cache to load vectors into memory for fast search), DiskBBQ dramatically reduces RAM requirements by avoiding the need to cache vectors in memory. This enables vector search on much larger datasets with minimal memory, though queries will be slower compared to in-memory approaches.
```

> Other quantization options include `int8_hnsw` (8-bit integer quantization) and `int4_hnsw` (4-bit integer quantization):

Reviewer 2: Let's add a subheading here describing the example

```console
PUT semantic-embeddings-int8
{
  "mappings": {
    "properties": {
      "content": {
        "type": "semantic_text",
        "inference_id": ".multilingual-e5-small-elasticsearch",
        "index_options": {
          "dense_vector": {
            "type": "int8_hnsw" <1>
          }
        }
      }
    }
  }
}
```

1. Use 8-bit integer quantization for 4x memory reduction with high accuracy retention. For 4-bit quantization, use `"type": "int4_hnsw"` instead, which provides up to 8x memory reduction. For the full list of other available quantization options (including `int4_flat` and others), refer to the [`dense_vector` `index_options` documentation](elasticsearch://reference/elasticsearch/mapping-reference/dense-vector.md#dense-vector-index-options).

> For HNSW-specific tuning parameters like `m` and `ef_construction`, you can include them in the `index_options`:

Reviewer 2: Let's add a subheading here describing the example
